File: ml-benchmark/autolifter/single-pass/line_sight.f
Config ExtraGrammar = "AutoLifter";

Inductive List = cons {Int, List} | nil Unit;

single_pass = \v: List -> Bool.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

inf = 100;

/*User provided programs*/

line_sight = fix (
  \f: Int -> List -> Bool. \ma: Int. \xs: List.
  match xs with
    nil _ -> true
  | cons {h, nil _} -> >= h ma
  | cons {h, t} ->
    if > h ma then f h t else f ma t
  end
) (- 0 inf);

main = single_pass line_sight;

File: ml-benchmark/autolifter/single-pass/line_sight.ml
type l =
  | Cons of int * l
  | Nil

let single_pass v =
  let rec run xs =
    match xs with
    | Nil -> xs
    | Cons (h, t) -> Cons (h, run t)
  in
  fun xs -> v (run xs)

let inf = 100

let rec line_sight_rec ma xs =
  match xs with
  | Nil -> true
  | Cons (h, t) ->
      match t with
      | Nil -> h >= ma
      | _ ->
          if h > ma then line_sight_rec h t else line_sight_rec ma t

let line_sight = line_sight_rec (0 - inf)

let program = single_pass line_sight